### PR TITLE
Bundle shared imports into the shell

### LINKED ISSUE #67: build: bundle shared resources with app shell
This removes a round trip to load shared-bundle.html

### PR DESCRIPTION
**Pull request by justinfagnani**
Fixes #67

### CODE REVIEW

**Comment by justinfagnani** [comment]
@azakus PTAL

File: src/build/bundle.ts [hunk 1]
```diff
@@ -176,6 +190,45 @@ export class Bundler {
     });
   }
 
+  _addSharedImportsToShell(bundles: Map<string, string[]>): string {
+    console.assert(this.shell != null);
+    let shellDeps = bundles.get(this.shell)
+        .slice()
```
Comment (azakus) [inline src/build/bundle.ts]: `slice is unnecessary`

File: src/build/bundle.ts [hunk 2]
```diff
@@ -176,6 +190,45 @@ export class Bundler {
...
+    let shellDepsSet = new Set(shellDeps);
+    let file = this.streamResolver._files.get(this.shell);
+    console.assert(file != null);
+    let contents = file.contents.toString();
+    let doc = dom5.parse(contents);
+    let imports = dom5.queryAll(doc, dom5.predicates.AND(
+      dom5.predicates.hasTagName('link'),
+      dom5.predicates.hasAttrValue('rel', 'import')
+    ));
+    for (let _import of imports) {
```
Comment (azakus) [inline src/build/bundle.ts]: Comment on why you rip the imports out of the doc if they are in the shared deps bundle

File: src/build/bundle.ts [hunk 3]
```diff
@@ -176,6 +190,45 @@ export class Bundler {
...
+      if (shellDepsSet.has(dom5.getAttribute(_import, 'href'))) {
+        dom5.remove(_import);
+      }
+    }
+    let head = dom5.query(doc, dom5.predicates.hasTagName('head'));
+    if (!head) {
```
Comment (azakus) [inline src/build/bundle.ts]: parse5 always generates at least `<html><head></head><body></body></html>`, so this case is always false.

**Comment by azakus** [comment]
LGTM

File: src/build/bundle.ts (hunk 1 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: done

File: src/build/bundle.ts (hunk 3 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: done

File: src/build/bundle.ts (hunk 2 above)
Comment (justinfagnani) [inline src/build/bundle.ts]: done

**Pull request merged**